rimage/transform/parallel_projection.go

```diff
@@ -95,5 +97,149 @@ func (pp *ParallelProjection) PointCloudToRGBD(cloud pointcloud.PointCloud) (*ri
 
 // ImagePointTo3DPoint takes the 2D pixel point and assumes that it represents the X,Y coordinate in mm as well.
 func (pp *ParallelProjection) ImagePointTo3DPoint(pt image.Point, d rimage.Depth) (r3.Vector, error) {
-	return r3.Vector{float64(pt.X), float64(pt.Y), float64(d)}, nil
+	return r3.Vector{X: float64(pt.X), Y: float64(pt.Y), Z: float64(d)}, nil
+}
+
+// ParallelProjectionOntoXZWithRobotMarker allows the creation of a 2D projection of a pointcloud and robot
+// position onto the XZ plane.
+type ParallelProjectionOntoXZWithRobotMarker struct {
+	robotPose *spatialmath.Pose
+}
+
+const (
+	sigmaLevel       = 7
+	imageHeight      = 1080
+	imageWidth       = 1080
+	missThreshold    = 0.44
+	hitThreshold     = 0.53
+	voxelRadius      = 1
+	robotMarkerRatio = 5
+)
+
+// PointCloudToRGBD creates an image of a pointcloud in the XZ plane, scaling the points to a standard image
+// size. It will also add a red marker to the map to represent the location of the robot. The returned depthMap
+// is unused and so will always be nil.
+func (ppRM *ParallelProjectionOntoXZWithRobotMarker) PointCloudToRGBD(cloud pointcloud.PointCloud,
+) (*rimage.Image, *rimage.DepthMap, error) {
+	meta := cloud.MetaData()
+
+	var X, Z []float64
+	cloud.Iterate(0, 0, func(pt r3.Vector, data pointcloud.Data) bool {
+		X = append(X, pt.X)
+		Z = append(Z, pt.Z)
+		return true
+	})
+
+	// Calculate max and min range to be represented in the output image and, if needed, cropping it based on
+	// the mean and standard deviation of the X and Z coordinates
+	meanX, err := stats.Mean(X)
```

Comment:
[Note] current stats package only errors out when data doesn't exist. There are no checks for overflow so the only error that can be triggers is the first check of meanX for null data.

Reply:
Also it seems `stats.Mean` returns a float32 but we use it with float64 below.... are we sure it is safe to mix & match the two?

Reply:
Sorry, ignore this, I was mistaken. When I followed the github reference to this function it lead me to https://github.com/jeremyhyde-viam/rdk/blob/6ae61768421dcb8d5f6baedf67e1bd6fbc656545/ml/inference/tflite_metadata/NormalizationOptions.go#L87 which is not what you are using. My mistake.

Reply:
One question though (non blocking for this PR just for my own context): I see that the golang code is using float64 for pointcloud data (which is good IMO) but the slam libraries use C++ float (which correct me if I'm wrong is only 32 bit). Is that decision of floating point precision already made for us by cartogroper & orbslam or could we support 64 bit floating point in the C++ library wrapper code we wrote by just changing the types?
 
Do we for see 32 bit floating point from the slam lib code being a problem?

Reply:
So... it looks like the idiomatic way to detect under / overflows in go is to check for whether or not the response value is infinite or not.

https://pkg.go.dev/math#IsInf can allow us to check for this. `math.IsInf(<float>, 0)` returns true if float is a positive or negative infinity.

Should we maybe add checks for that condition & return an error to the end user if we over/underflow? 

If so, I think we can still use the stats package.

```golang
package main

import (
	"fmt"
	"math"
)

func main() {
  x := math.MaxFloat64
  x2 := x*2.0
  fmt.Printf("math.MaxFloat64: is infinity: %b\n", math.IsInf(x, 0))
  fmt.Printf("math.MaxFloat64*2: is infinity: %b\n", math.IsInf(x2, 0))
}
```

prints:
```
math.MaxFloat64: is infinity: %!b(bool=false)
math.MaxFloat64*2: is infinity: %!b(bool=true)
```